### ISSUE TITLE
Installation in pytorch==1.10.0 failure

### ISSUE THREAD

**Issue by Tengxu-Sun** [laekov/fastmoe#87]
I have a try on fastmoe under pytorch==1.10.0 but get failure, please have a look, thanks.

**Describe the bug**
Installation in pytorch==1.10.0 via `USE_NCCL=1 python setup.py install' failure

**To Reproduce**
1. install pytorch via `pip install torch==1.10.0+cu111 -f https://download.pytorch.org/whl/torch_stable.html`
2. install apex via  https://github.com/NVIDIA/apex#linux
2. Run `USE_NCCL=1 python setup.py install`  processes on "v100" 

**Expected behavior**
successs

**Logs**
```
running install
running bdist_egg
running egg_info
writing fastmoe.egg-info/PKG-INFO
writing dependency_links to fastmoe.egg-info/dependency_links.txt
writing top-level names to fastmoe.egg-info/top_level.txt
reading manifest file 'fastmoe.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'fastmoe.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
running build_ext
building 'fmoe_cuda' extension
Emitting ninja build file /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/build.ninja...
Compiling objects...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
[1/2] c++ -MMD -MF /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/global_exchange.o.d -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/lib/python3.7/site-packages/torch/include -I/opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include -I/opt/conda/lib/python3.7/site-packages/torch/include/TH -I/opt/conda/lib/python3.7/site-packages/torch/include/THC -I/usr/local/cuda/include -I/opt/conda/include/python3.7m -c -c /home/***/fastmoe_shen/cuda/global_exchange.cpp -o /home/suntengxu/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/global_exchange.o -DFMOE_USE_NCCL -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=fmoe_cuda -DTORCH_EXTENSION_NAME=fmoe_cuda -D_GLIBCXX_USE_CXX11_ABI=0 -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14
FAILED: /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/global_exchange.o 
c++ -MMD -MF /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/global_exchange.o.d -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/lib/python3.7/site-packages/torch/include -I/opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include -I/opt/conda/lib/python3.7/site-packages/torch/include/TH -I/opt/conda/lib/python3.7/site-packages/torch/include/THC -I/usr/local/cuda/include -I/opt/conda/include/python3.7m -c -c /home/***/fastmoe_shen/cuda/global_exchange.cpp -o /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/global_exchange.o -DFMOE_USE_NCCL -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=fmoe_cuda -DTORCH_EXTENSION_NAME=fmoe_cuda -D_GLIBCXX_USE_CXX11_ABI=0 -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
/home/***/fastmoe_shen/cuda/global_exchange.cpp:76:29: error: ‘c10d’ has not been declared
 class HackNCCLGroup: public c10d::ProcessGroupNCCL {
                             ^~~~
/home/***/fastmoe_shen/cuda/global_exchange.cpp:76:35: error: expected ‘{’ before ‘ProcessGroupNCCL’
 class HackNCCLGroup: public c10d::ProcessGroupNCCL {
                                   ^~~~~~~~~~~~~~~~
/home/***/fastmoe_shen/cuda/global_exchange.cpp:77:1: error: expected primary-expression before ‘public’
 public:
 ^~~~~~
/home/***/fastmoe_shen/cuda/global_exchange.cpp:77:1: error: expected ‘}’ before ‘public’
/home/***/fastmoe_shen/cuda/global_exchange.cpp:97:1: error: expected declaration before ‘}’ token
 };
 ^
[2/2] c++ -MMD -MF /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/fmoe_cuda.o.d -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/lib/python3.7/site-packages/torch/include -I/opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include -I/opt/conda/lib/python3.7/site-packages/torch/include/TH -I/opt/conda/lib/python3.7/site-packages/torch/include/THC -I/usr/local/cuda/include -I/opt/conda/include/python3.7m -c -c /home/***/fastmoe_shen/cuda/fmoe_cuda.cpp -o /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/fmoe_cuda.o -DFMOE_USE_NCCL -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=fmoe_cuda -DTORCH_EXTENSION_NAME=fmoe_cuda -D_GLIBCXX_USE_CXX11_ABI=0 -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14
FAILED: /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/fmoe_cuda.o 
c++ -MMD -MF /home/suntengxu/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/fmoe_cuda.o.d -pthread -B /opt/conda/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/opt/conda/lib/python3.7/site-packages/torch/include -I/opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include -I/opt/conda/lib/python3.7/site-packages/torch/include/TH -I/opt/conda/lib/python3.7/site-packages/torch/include/THC -I/usr/local/cuda/include -I/opt/conda/include/python3.7m -c -c /home/***/fastmoe_shen/cuda/fmoe_cuda.cpp -o /home/***/fastmoe_shen/build/temp.linux-x86_64-3.7/cuda/fmoe_cuda.o -DFMOE_USE_NCCL -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' '-DPYBIND11_BUILD_ABI="_cxxabi1011"' -DTORCH_EXTENSION_NAME=fmoe_cuda -DTORCH_EXTENSION_NAME=fmoe_cuda -D_GLIBCXX_USE_CXX11_ABI=0 -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
/home/***/fastmoe_shen/cuda/fmoe_cuda.cpp:21:19: error: variable or field ‘_ensure_nccl’ declared void
 void _ensure_nccl(c10d::ProcessGroupNCCL& p, torch::Tensor t);
                   ^~~~
/home/***/fastmoe_shen/cuda/fmoe_cuda.cpp:21:19: error: ‘c10d’ has not been declared
/home/***/fastmoe_shen/cuda/fmoe_cuda.cpp:21:43: error: ‘p’ was not declared in this scope
 void _ensure_nccl(c10d::ProcessGroupNCCL& p, torch::Tensor t);
                                           ^
/home/***/fastmoe_shen/cuda/fmoe_cuda.cpp:21:43: note: suggested alternatives:
In file included from /opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/Dimname.h:3:0,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/NamedTensor.h:3,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/TensorBody.h:24,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/ATen/Tensor.h:3,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/ATen/Context.h:4,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/ATen/ATen.h:9,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include/torch/types.h:3,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include/torch/data/dataloader_options.h:4,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include/torch/data/dataloader/base.h:3,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include/torch/data/dataloader/stateful.h:3,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include/torch/data/dataloader.h:3,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include/torch/data.h:3,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/torch/csrc/api/include/torch/all.h:8,
                 from /opt/conda/lib/python3.7/site-packages/torch/include/torch/extension.h:4,
                 from /home/suntengxu/fastmoe_shen/cuda/fmoe_cuda.cpp:3:
/opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/aten_interned_strings.h:962:9: note:   ‘c10::attr::p’
 _(attr, p) \
         ^
/opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/interned_strings.h:615:35: note: in definition of macro ‘DEFINE_SYMBOL’
   namespace ns { constexpr Symbol s(static_cast<unique_t>(_keys::ns##_##s)); }
                                   ^
/opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/interned_strings.h:448:3: note: in expansion of macro ‘FORALL_ATTR_BASE_SYMBOLS’
   FORALL_ATTR_BASE_SYMBOLS(_)        \
   ^~~~~~~~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/interned_strings.h:616:1: note: in expansion of macro ‘FORALL_NS_SYMBOLS’
 FORALL_NS_SYMBOLS(DEFINE_SYMBOL)
 ^~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/aten_interned_strings.h:962:9: note:   ‘c10::attr::p’
 _(attr, p) \
         ^
/opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/interned_strings.h:615:35: note: in definition of macro ‘DEFINE_SYMBOL’
   namespace ns { constexpr Symbol s(static_cast<unique_t>(_keys::ns##_##s)); }
                                   ^
/opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/interned_strings.h:448:3: note: in expansion of macro ‘FORALL_ATTR_BASE_SYMBOLS’
   FORALL_ATTR_BASE_SYMBOLS(_)        \
   ^~~~~~~~~~~~~~~~~~~~~~~~
/opt/conda/lib/python3.7/site-packages/torch/include/ATen/core/interned_strings.h:616:1: note: in expansion of macro ‘FORALL_NS_SYMBOLS’
 FORALL_NS_SYMBOLS(DEFINE_SYMBOL)
 ^~~~~~~~~~~~~~~~~
/home/suntengxu/fastmoe_shen/cuda/fmoe_cuda.cpp:21:60: error: expected primary-expression before ‘t’
 void _ensure_nccl(c10d::ProcessGroupNCCL& p, torch::Tensor t);
                                                            ^
/home/suntengxu/fastmoe_shen/cuda/fmoe_cuda.cpp: In function ‘void pybind11_init_fmoe_cuda(pybind11::module_&)’:
/home/suntengxu/fastmoe_shen/cuda/fmoe_cuda.cpp:61:27: error: ‘_ensure_nccl’ was not declared in this scope
     m.def("ensure_nccl", &_ensure_nccl, "FastMoE ensure torch nccl comm");
                           ^~~~~~~~~~~~
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1723, in _run_ninja_build
    env=env)
  File "/opt/conda/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "setup.py", line 52, in <module>
    'build_ext': BuildExtension
  File "/opt/conda/lib/python3.7/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/opt/conda/lib/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/opt/conda/lib/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/opt/conda/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/opt/conda/lib/python3.7/site-packages/setuptools/command/install.py", line 67, in run
    self.do_egg_install()
  File "/opt/conda/lib/python3.7/site-packages/setuptools/command/install.py", line 109, in do_egg_install
    self.run_command('bdist_egg')
  File "/opt/conda/lib/python3.7/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/opt/conda/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/opt/conda/lib/python3.7/site-packages/setuptools/command/bdist_egg.py", line 164, in run
    cmd = self.call_command('install_lib', warn_dir=0)
  File "/opt/conda/lib/python3.7/site-packages/setuptools/command/bdist_egg.py", line 150, in call_command
    self.run_command(cmdname)
  File "/opt/conda/lib/python3.7/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/opt/conda/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/opt/conda/lib/python3.7/site-packages/setuptools/command/install_lib.py", line 11, in run
    self.build()
  File "/opt/conda/lib/python3.7/distutils/command/install_lib.py", line 107, in build
    self.run_command('build_ext')
  File "/opt/conda/lib/python3.7/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/opt/conda/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/opt/conda/lib/python3.7/site-packages/setuptools/command/build_ext.py", line 79, in run
    _build_ext.run(self)
  File "/opt/conda/lib/python3.7/distutils/command/build_ext.py", line 340, in run
    self.build_extensions()
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 735, in build_extensions
    build_ext.build_extensions(self)
  File "/opt/conda/lib/python3.7/distutils/command/build_ext.py", line 449, in build_extensions
    self._build_extensions_serial()
  File "/opt/conda/lib/python3.7/distutils/command/build_ext.py", line 474, in _build_extensions_serial
    self.build_extension(ext)
  File "/opt/conda/lib/python3.7/site-packages/setuptools/command/build_ext.py", line 196, in build_extension
    _build_ext.build_extension(self, ext)
  File "/opt/conda/lib/python3.7/distutils/command/build_ext.py", line 534, in build_extension
    depends=ext.depends)
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 565, in unix_wrap_ninja_compile
    with_cuda=with_cuda)
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1404, in _write_ninja_file_and_compile_objects
    error_prefix='Error compiling objects for extension')
  File "/opt/conda/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1733, in _run_ninja_build
    raise RuntimeError(message) from e
RuntimeError: Error compiling objects for extension
```

**Platform**
 - Device: [NVIDIA V100]
 - OS: [Ubuntu 18.04.5 LTS]
 - CUDA version: [ 11.1]
 - NCCL version: [e.g. 2.7.8-1]
 - PyTorch version: [1.10.0]
 - gcc: [7.5.0]
 - ninja version: [ 1.9.0/1.10.2]

**Additional context**

it seems like some code error as shown in 
```
/home/***/fastmoe_shen/cuda/global_exchange.cpp:76:29: error: ‘c10d’ has not been declared
 class HackNCCLGroup: public c10d::ProcessGroupNCCL {
                             ^~~~
/home/***/fastmoe_shen/cuda/global_exchange.cpp:76:35: error: expected ‘{’ before ‘ProcessGroupNCCL’
 class HackNCCLGroup: public c10d::ProcessGroupNCCL {
                                   ^~~~~~~~~~~~~~~~
/home/***/fastmoe_shen/cuda/global_exchange.cpp:77:1: error: expected primary-expression before ‘public’
 public:
 ^~~~~~
/home/***/fastmoe_shen/cuda/global_exchange.cpp:77:1: error: expected ‘}’ before ‘public’
/home/***/fastmoe_shen/cuda/global_exchange.cpp:97:1: error: expected declaration before ‘}’ token
 };
```
```
/home/***/fastmoe_shen/cuda/fmoe_cuda.cpp:21:19: error: variable or field ‘_ensure_nccl’ declared void
 void _ensure_nccl(c10d::ProcessGroupNCCL& p, torch::Tensor t);
                   ^~~~
/home/***/fastmoe_shen/cuda/fmoe_cuda.cpp:21:19: error: ‘c10d’ has not been declared
/home/***/fastmoe_shen/cuda/fmoe_cuda.cpp:21:43: error: ‘p’ was not declared in this scope
 void _ensure_nccl(c10d::ProcessGroupNCCL& p, torch::Tensor t);
```
but can get success under pytroch=1.9.0. This is very weird. Can it be a version problem？
Any idea on how to solved this would be grateful.

**Comment by Tengxu-Sun** [comment]
Some files did change their place https://github.com/pytorch/pytorch/pull/60543 under torch==1.10.0, but still in the same path after installation.  The core usage also has not changed.

**Comment by laekov** [comment]
Fixed in #88
